###### Study Project - Team members : Chandrakanth Yela, Ishan Agarwal, Aravind Reddy Keesara

# Importing packages

In [950]:
import pandas as pd
import numpy as np
import time

# Reading Benign Data

In [951]:
db1 = pd.read_csv('benign_2013_1.csv')
db1['year']='January 2013'
db2 = pd.read_csv('benign_2013_2.csv')
db2['year']='February 2013'
db3 = pd.read_csv('benign_2013_3.csv')
db3['year']='March 2013'
db4 = pd.read_csv('benign_2013_4.csv')
db4['year']='April 2013'
db5 = pd.read_csv('benign_2013_5.csv')
db5['year']='May 2013'
db6 = pd.read_csv('benign_2013_6.csv')
db6['year']='June 2013'
db7 = pd.read_csv('benign_2013_7.csv')
db7['year']='July 2013'
db8 = pd.read_csv('benign_2013_8.csv')
db8['year']='August 2013'
db9 = pd.read_csv('benign_2013_9.csv')
db9['year']='September 2013'
db10 = pd.read_csv('benign_2013_10.csv')
db10['year']='October 2013'
db11 = pd.read_csv('benign_2013_11.csv')
db11['year']='November 2013'
db12 = pd.read_csv('benign_2013_12.csv')
db12['year']='December 2013'
db13 = pd.read_csv('benign_2014_1.csv')
db13['year']='January 2014'
db14 = pd.read_csv('benign_2014_2.csv')
db14['year']='February 2014'
db15 = pd.read_csv('benign_2014_3.csv')
db15['year']='March 2014'
db16 = pd.read_csv('benign_2014_4.csv')
db16['year']='April 2014'
db17 = pd.read_csv('benign_2014_5.csv')
db17['year']='May 2014'
db18 = pd.read_csv('benign_2014_6.csv')
db18['year']='June 2014'
db19 = pd.read_csv('benign_2014_7.csv')
db19['year']='July 2014'

# Merging the benign dataset

In [952]:
combined = db1
combined = combined.append([db2,db3,db4,db5,db6,db7,db8,db9,db10,db11,db12,db13,db14,db15,db16,db17,db18,db19])
combined=pd.DataFrame(combined)
combined.head()

username      revid               revtime  \
0  Swiftscw  535365129  2013-01-28T16:43:56Z   
1  Swiftscw  535365309  2013-01-28T16:45:08Z   
2  Swiftscw  535366071  2013-01-28T16:50:17Z   
3  Swiftscw  535369287  2013-01-28T17:13:40Z   
4  Swiftscw  535370277  2013-01-28T17:21:30Z   

                           pagetitle  isReverted revertTime  cluebotRevert  \
0                      User:Swiftscw       False          -              0   
1                      User:Swiftscw       False          -              0   
2                      User:Swiftscw       False          -              0   
3       User:Swiftscw/AMAPOLA CABASE       False          -              0   
4  Wikipedia:Requests for undeletion       False          -              0   

   stiki_score  stiki_REP_USER          year  
0          0.0             0.0  January 2013  
1          0.0             0.0  January 2013  
2          0.0             0.0  January 2013  
3          0.0             0.0  January 2013  
4          0.0             0.0  January 2013

# Reading user file

In [953]:
colnames = ['username','userid','blocked_time','blocked_reason','type']
type(colnames)
users=pd.read_table('users.tsv',sep="\t",header=None,names=colnames)
users=pd.DataFrame(users)
users.head()

username    userid blocked_time blocked_reason    type
0      Swiftscw  18342896            -              -  benign
1         Azkle  18359941            -              -  benign
2    Tgrosinger  18228795            -              -  benign
3  BanglarBogra  18197517            -              -  benign
4     Tiny Trot  18315964            -              -  benign


# Benign and User Left Join on "username"

In [954]:
ben = pd.merge(combined,users,how='left',on='username')
ben.head(10)

username      revid               revtime  \
0  Swiftscw  535365129  2013-01-28T16:43:56Z   
1  Swiftscw  535365309  2013-01-28T16:45:08Z   
2  Swiftscw  535366071  2013-01-28T16:50:17Z   
3  Swiftscw  535369287  2013-01-28T17:13:40Z   
4  Swiftscw  535370277  2013-01-28T17:21:30Z   
5  Swiftscw  535376382  2013-01-28T18:07:16Z   
6  Swiftscw  535379254  2013-01-28T18:27:12Z   
7  Swiftscw  535380411  2013-01-28T18:34:51Z   
8  Swiftscw  535382439  2013-01-28T18:48:17Z   
9  Swiftscw  535390355  2013-01-28T19:40:51Z   

                           pagetitle  isReverted revertTime  cluebotRevert  \
0                      User:Swiftscw       False          -              0   
1                      User:Swiftscw       False          -              0   
2                      User:Swiftscw       False          -              0   
3       User:Swiftscw/AMAPOLA CABASE       False          -              0   
4  Wikipedia:Requests for undeletion       False          -              0   
5                       Maria Cabase       False          -              0   
6                       Maria Cabase       False          -              0   
7                       Maria Cabase       False          -              0   
8                       Maria Cabase       False          -              0   
9                       Maria Cabase       False          -              0   

   stiki_score  stiki_REP_USER          year    userid blocked_time  \
0     0.000000             0.0  January 2013  18342896            -   
1     0.000000             0.0  January 2013  18342896            -   
2     0.000000             0.0  January 2013  18342896            -   
3     0.000000             0.0  January 2013  18342896            -   
4     0.000000             0.0  January 2013  18342896            -   
5     0.292556             0.0  January 2013  18342896            -   
6     0.282316             0.0  January 2013  18342896            -   
7     0.367025             0.0  January 2013  18342896            -   
8     0.282316             0.0  January 2013  18342896            -   
9     0.132504             0.0  January 2013  18342896            -   

  blocked_reason    type  
0              -  benign  
1              -  benign  
2              -  benign  
3              -  benign  
4              -  benign  
5              -  benign  
6              -  benign  
7              -  benign  
8              -  benign  
9              -  benign

# Reading Vandal data

In [955]:
dv1 = pd.read_csv('vandal_2013_1.csv')
dv1['year']='January 2013'
dv2 = pd.read_csv('vandal_2013_2.csv')
dv2['year']='February 2013'
dv3 = pd.read_csv('vandal_2013_3.csv')
dv3['year']='March 2013'
dv4 = pd.read_csv('vandal_2013_4.csv')
dv4['year']='April 2013'
dv5 = pd.read_csv('vandal_2013_5.csv')
dv5['year']='May 2013'
dv6 = pd.read_csv('vandal_2013_6.csv')
dv6['year']='June 2013'
dv7 = pd.read_csv('vandal_2013_7.csv')
dv7['year']='July 2013'
dv8 = pd.read_csv('vandal_2013_8.csv')
dv8['year']='August 2013'
dv9 = pd.read_csv('vandal_2013_9.csv')
dv9['year']='September 2013'
dv10 = pd.read_csv('vandal_2013_10.csv')
dv10['year']='October 2013'
dv11 = pd.read_csv('vandal_2013_11.csv')
dv11['year']='November 2013'
dv12 = pd.read_csv('vandal_2013_12.csv')
dv12['year']='December 2013'
dv13 = pd.read_csv('vandal_2014_1.csv')
dv13['year']='January 2014'
dv14 = pd.read_csv('vandal_2014_2.csv')
dv14['year']='February 2014'
dv15 = pd.read_csv('vandal_2014_3.csv')
dv15['year']='March 2014'
dv16 = pd.read_csv('vandal_2014_4.csv')
dv16['year']='April 2014'
dv17 = pd.read_csv('vandal_2014_5.csv')
dv17['year']='May 2014'
dv18 = pd.read_csv('vandal_2014_6.csv')
dv18['year']='June 2014'
dv19 = pd.read_csv('vandal_2014_7.csv')
dv19['year']='July 2014'

# Merging the Vandal dataset

In [956]:
comvandal = dv1
comvandal = comvandal.append([dv2,dv3,dv4,dv5,dv6,dv7,dv8,dv9,dv10,dv11,dv12,dv13,dv14,dv15,dv16,dv17,dv18,dv19])
comvandal=pd.DataFrame(comvandal)
comvandal.head()

username      revid               revtime      pagetitle  isReverted  \
0  Alex12rocks  535937612  2013-01-31T23:19:17Z  Pueblo Revolt        True   
1  Alex12rocks  535937732  2013-01-31T23:20:04Z  Pueblo Revolt        True   
2  Alex12rocks  535937933  2013-01-31T23:21:19Z  Pueblo Revolt        True   
3  Alex12rocks  535938394  2013-01-31T23:24:08Z  Pueblo Revolt        True   
4  Alex12rocks  535938540  2013-01-31T23:24:56Z  Pueblo Revolt        True   

             revertTime  cluebotRevert  stiki_score  stiki_REP_USER  \
0  2013-01-31T23:20:35Z              0     0.462073        0.000000   
1  2013-01-31T23:20:35Z              0     0.437085        0.000000   
2  2013-01-31T23:21:30Z              0     0.648625        0.999851   
3  2013-01-31T23:24:15Z              0     0.439547        1.999305   
4  2013-01-31T23:24:59Z              1     0.618692        2.998948   

           year  
0  January 2013  
1  January 2013  
2  January 2013  
3  January 2013  
4  January 2013

# Vandal and User Left Join on "username"

In [957]:
van = pd.merge(comvandal,users,how='left',on='username')
van.head()

username      revid               revtime      pagetitle  isReverted  \
0  Alex12rocks  535937612  2013-01-31T23:19:17Z  Pueblo Revolt        True   
1  Alex12rocks  535937732  2013-01-31T23:20:04Z  Pueblo Revolt        True   
2  Alex12rocks  535937933  2013-01-31T23:21:19Z  Pueblo Revolt        True   
3  Alex12rocks  535938394  2013-01-31T23:24:08Z  Pueblo Revolt        True   
4  Alex12rocks  535938540  2013-01-31T23:24:56Z  Pueblo Revolt        True   

             revertTime  cluebotRevert  stiki_score  stiki_REP_USER  \
0  2013-01-31T23:20:35Z              0     0.462073        0.000000   
1  2013-01-31T23:20:35Z              0     0.437085        0.000000   
2  2013-01-31T23:21:30Z              0     0.648625        0.999851   
3  2013-01-31T23:24:15Z              0     0.439547        1.999305   
4  2013-01-31T23:24:59Z              1     0.618692        2.998948   

           year   userid          blocked_time  \
0  January 2013  4232100  2013-01-31T23:34:10Z   
1  January 2013  4232100  2013-01-31T23:34:10Z   
2  January 2013  4232100  2013-01-31T23:34:10Z   
3  January 2013  4232100  2013-01-31T23:34:10Z   
4  January 2013  4232100  2013-01-31T23:34:10Z   

                                      blocked_reason    type  
0  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  
1  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  
2  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  
3  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  
4  [[WP:Vandalism-only account|Vandalism-only acc...  vandal


# Reading Page Details Data

In [958]:
pages=pd.read_csv('pages.tsv')
pages=pd.DataFrame(pages)
pages.head(10)

pagetitle    pageid  \
0                            User:Swiftscw  38354683   
1             User:Swiftscw/AMAPOLA CABASE  38354972   
2        Wikipedia:Requests for undeletion  26803395   
3                             Maria Cabase  32754928   
4                        Talk:Maria Cabase  34549656   
5                    Manuel Kabajar Cabase  32755116   
6              User talk:Postdlf/Archive23  38090775   
7  Wikipedia:WikiProject Biography/Members   6163553   
8                       User talk:Swiftscw  40011085   
9                           Bobby Enriquez   7040405   

                                      pagecategories  
0  set([u'Category:WikiProject Biography particip...  
1                                            set([])  
2  set([u'Category:Wikipedia maintenance', u'Cate...  
3  set([u'Category:Filipino pianists', u'Category...  
4  set([u'Category:WikiProject Biography articles...  
5  set([u'Category:Filipino drummers', u'Category...  
6                                            set([])  
7                                            set([])  
8                                            set([])  
9  set([u'Category:Filipino Pentecostals', u'Cate...

# Left Join Vandal and Benign with Page data on "pagetitle"

In [959]:
van_final = pd.merge(van,pages,how='left',on="pagetitle")
ben_final = pd.merge(ben,pages,how='left',on="pagetitle")
van_final.head(10)
ben_final.head(10)

username      revid               revtime  \
0  Swiftscw  535365129  2013-01-28T16:43:56Z   
1  Swiftscw  535365309  2013-01-28T16:45:08Z   
2  Swiftscw  535366071  2013-01-28T16:50:17Z   
3  Swiftscw  535369287  2013-01-28T17:13:40Z   
4  Swiftscw  535370277  2013-01-28T17:21:30Z   
5  Swiftscw  535370277  2013-01-28T17:21:30Z   
6  Swiftscw  535376382  2013-01-28T18:07:16Z   
7  Swiftscw  535379254  2013-01-28T18:27:12Z   
8  Swiftscw  535380411  2013-01-28T18:34:51Z   
9  Swiftscw  535382439  2013-01-28T18:48:17Z   

                           pagetitle  isReverted revertTime  cluebotRevert  \
0                      User:Swiftscw       False          -              0   
1                      User:Swiftscw       False          -              0   
2                      User:Swiftscw       False          -              0   
3       User:Swiftscw/AMAPOLA CABASE       False          -              0   
4  Wikipedia:Requests for undeletion       False          -              0   
5  Wikipedia:Requests for undeletion       False          -              0   
6                       Maria Cabase       False          -              0   
7                       Maria Cabase       False          -              0   
8                       Maria Cabase       False          -              0   
9                       Maria Cabase       False          -              0   

   stiki_score  stiki_REP_USER          year    userid blocked_time  \
0     0.000000             0.0  January 2013  18342896            -   
1     0.000000             0.0  January 2013  18342896            -   
2     0.000000             0.0  January 2013  18342896            -   
3     0.000000             0.0  January 2013  18342896            -   
4     0.000000             0.0  January 2013  18342896            -   
5     0.000000             0.0  January 2013  18342896            -   
6     0.292556             0.0  January 2013  18342896            -   
7     0.282316             0.0  January 2013  18342896            -   
8     0.367025             0.0  January 2013  18342896            -   
9     0.282316             0.0  January 2013  18342896            -   

  blocked_reason    type    pageid  \
0              -  benign  38354683   
1              -  benign  38354683   
2              -  benign  38354683   
3              -  benign  38354972   
4              -  benign  26803395   
5              -  benign  26803395   
6              -  benign  32754928   
7              -  benign  32754928   
8              -  benign  32754928   
9              -  benign  32754928   

                                      pagecategories  
0  set([u'Category:WikiProject Biography particip...  
1  set([u'Category:WikiProject Biography particip...  
2  set([u'Category:WikiProject Biography particip...  
3                                            set([])  
4  set([u'Category:Wikipedia maintenance', u'Cate...  
5  set([u'Category:Wikipedia maintenance', u'Cate...  
6  set([u'Category:Filipino pianists', u'Category...  
7  set([u'Category:Filipino pianists', u'Category...  
8  set([u'Category:Filipino pianists', u'Category...  
9  set([u'Category:Filipino pianists', u'Category...

# Cleaning the data 

In [960]:
import pandas as pd

In [961]:
van_final.apply(lambda x: sum(x.isna()))

username            0
revid               0
revtime             0
pagetitle           0
isReverted          0
revertTime          0
cluebotRevert       0
stiki_score         0
stiki_REP_USER    727
year                0
userid              0
blocked_time        0
blocked_reason      0
type                0
pageid              0
pagecategories      0
dtype: int64

In [962]:

users_van = van_final.dropna(axis=0)
users_van.apply(lambda x: sum(x.isna()))

username          0
revid             0
revtime           0
pagetitle         0
isReverted        0
revertTime        0
cluebotRevert     0
stiki_score       0
stiki_REP_USER    0
year              0
userid            0
blocked_time      0
blocked_reason    0
type              0
pageid            0
pagecategories    0
dtype: int64

In [963]:

users_ben = ben_final.dropna(axis=0)
users_ben.apply(lambda x: sum(x.isna()))

username          0
revid             0
revtime           0
pagetitle         0
isReverted        0
revertTime        0
cluebotRevert     0
stiki_score       0
stiki_REP_USER    0
year              0
userid            0
blocked_time      0
blocked_reason    0
type              0
pageid            0
pagecategories    0
dtype: int64

# Merging Vandal and Benign Data

In [964]:
users_fin=users_van.append(users_ben)
users_fin.head()

username      revid               revtime      pagetitle  isReverted  \
0  Alex12rocks  535937612  2013-01-31T23:19:17Z  Pueblo Revolt        True   
1  Alex12rocks  535937732  2013-01-31T23:20:04Z  Pueblo Revolt        True   
2  Alex12rocks  535937933  2013-01-31T23:21:19Z  Pueblo Revolt        True   
3  Alex12rocks  535938394  2013-01-31T23:24:08Z  Pueblo Revolt        True   
4  Alex12rocks  535938540  2013-01-31T23:24:56Z  Pueblo Revolt        True   

             revertTime  cluebotRevert  stiki_score  stiki_REP_USER  \
0  2013-01-31T23:20:35Z              0     0.462073        0.000000   
1  2013-01-31T23:20:35Z              0     0.437085        0.000000   
2  2013-01-31T23:21:30Z              0     0.648625        0.999851   
3  2013-01-31T23:24:15Z              0     0.439547        1.999305   
4  2013-01-31T23:24:59Z              1     0.618692        2.998948   

           year   userid          blocked_time  \
0  January 2013  4232100  2013-01-31T23:34:10Z   
1  January 2013  4232100  2013-01-31T23:34:10Z   
2  January 2013  4232100  2013-01-31T23:34:10Z   
3  January 2013  4232100  2013-01-31T23:34:10Z   
4  January 2013  4232100  2013-01-31T23:34:10Z   

                                      blocked_reason    type  pageid  \
0  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  569572   
1  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  569572   
2  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  569572   
3  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  569572   
4  [[WP:Vandalism-only account|Vandalism-only acc...  vandal  569572   

                                      pagecategories  
0  set([u'Category:17th century in New Spain', u'...  
1  set([u'Category:17th century in New Spain', u'...  
2  set([u'Category:17th century in New Spain', u'...  
3  set([u'Category:17th century in New Spain', u'...  
4  set([u'Category:17th century in New Spain', u'...

### converting date to proper format

In [965]:
ben_final.revtime = ben_final.revtime.str[:19]


In [966]:
ben_final.shape

(657863, 16)

In [967]:
van.revtime = ben_final.revtime.str[:19]

In [968]:
ben_final['revtime'] = pd.to_datetime(ben_final['revtime'])

In [969]:
van_final['revtime'] = pd.to_datetime(van_final['revtime'])

##### seperating some cloumns to process other features in the later part of analysis

In [970]:
ben_scores = ben_final.loc[:,['userid','stiki_score','cluebotRevert']]
van_scores = van_final.loc[:,['userid','stiki_score','cluebotRevert']]

###   1)  checking whether the first edit of user is a metapage

In [971]:
ben_final[ben_final['pagetitle'].str.contains('/')]

username      revid             revtime  \
3                     Swiftscw  535369287 2013-01-28 17:13:40   
16                    Swiftscw  535402422 2013-01-28 21:00:26   
18                    Swiftscw  564933523 2013-07-19 15:02:39   
26                       Azkle  535906972 2013-01-31 19:50:17   
97                  Tgrosinger  537167961 2013-02-08 05:00:23   
100                 Tgrosinger  539104591 2013-02-19 20:42:51   
106                 Tgrosinger  539914351 2013-02-23 16:49:06   
118                 Tgrosinger  540982409 2013-02-27 16:42:58   
119                 Tgrosinger  540989343 2013-02-27 17:12:00   
152               Naveenchand3  531288293 2013-01-04 15:55:16   
158                    Griff88  533205575 2013-01-15 14:29:37   
159                    Griff88  533206041 2013-01-15 14:34:15   
160                    Griff88  533343359 2013-01-16 09:46:16   
161                    Griff88  533343583 2013-01-16 09:49:54   
176                    Griff88  550442023 2013-04-15 08:22:00   
186                    Griff88  556398136 2013-05-23 08:49:05   
268                    Griff88  587348043 2013-12-23 08:38:58   
271                    Griff88  587352363 2013-12-23 09:33:19   
322                    Griff88  598223958 2014-03-05 07:04:43   
329                    Griff88  610605412 2014-05-29 07:30:19   
330                    Griff88  610612755 2014-05-29 08:57:23   
368                    Griff88  613258684 2014-06-17 09:10:17   
418                    Griff88  620346665 2014-08-08 08:01:05   
472                    Griff88  623962199 2014-09-03 05:14:30   
477                    Griff88  623994300 2014-09-03 11:09:25   
584             DilbertReality  563757846 2013-07-11 02:33:21   
661                Laurenwking  536558332 2013-02-04 18:41:28   
662                Laurenwking  536559529 2013-02-04 18:48:48   
663                Laurenwking  536562451 2013-02-04 19:07:23   
670                Laurenwking  539218740 2013-02-20 12:25:29   
...                        ...        ...                 ...   
657513                Ddosguru  620405660 2014-08-08 18:21:42   
657514                Ddosguru  620405714 2014-08-08 18:22:08   
657529                Ddosguru  657220490 2015-04-19 18:59:45   
657530                Ddosguru  657220897 2015-04-19 19:03:01   
657531                Ddosguru  657221862 2015-04-19 19:11:34   
657697                 Isanaht  619902707 2014-08-05 02:12:59   
657698                 Isanaht  619902799 2014-08-05 02:14:17   
657773             Africulture  623454973 2014-08-30 15:03:48   
657774             Africulture  623454973 2014-08-30 15:03:48   
657800  Nothingknewunderthesun  625154022 2014-09-11 22:14:01   
657801  Nothingknewunderthesun  625155247 2014-09-11 22:24:24   
657802  Nothingknewunderthesun  625157768 2014-09-11 22:45:04   
657803  Nothingknewunderthesun  625157768 2014-09-11 22:45:04   
657804  Nothingknewunderthesun  625863565 2014-09-16 21:02:41   
657805  Nothingknewunderthesun  625863655 2014-09-16 21:03:33   
657806  Nothingknewunderthesun  625863894 2014-09-16 21:05:43   
657807  Nothingknewunderthesun  625864880 2014-09-16 21:14:30   
657808  Nothingknewunderthesun  625865326 2014-09-16 21:17:59   
657809  Nothingknewunderthesun  625868880 2014-09-16 21:48:03   
657810  Nothingknewunderthesun  625869984 2014-09-16 21:58:29   
657811  Nothingknewunderthesun  625870858 2014-09-16 22:06:55   
657812  Nothingknewunderthesun  625877618 2014-09-16 23:21:29   
657827           KhalidAliHaji  626880826 2014-09-24 10:41:04   
657828           KhalidAliHaji  626882376 2014-09-24 11:03:17   
657829           KhalidAliHaji  626883014 2014-09-24 11:11:57   
657830           KhalidAliHaji  626883112 2014-09-24 11:13:17   
657831           KhalidAliHaji  626883191 2014-09-24 11:14:19   
657832           KhalidAliHaji  626883293 2014-09-24 11:15:25   
657833           KhalidAliHaji  626883541 2014-09-24 11:18:41   
657857               Dinkjunky  621772362 2014-08-18 13:4

In [972]:
van_final[ben_final['pagetitle'].str.contains('/')]

/Users/owner/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


username      revid             revtime  \
3           Alex12rocks  535938394 2013-01-31 23:24:08   
16         RevengeOfBHS  535045160 2013-01-26 21:06:32   
18         RevengeOfBHS  535045725 2013-01-26 21:09:41   
26       Sultymcbuttsex  535930942 2013-01-31 22:33:28   
97      Johnny89detroit  535907433 2013-01-31 19:53:39   
100     Johnny89detroit  535908105 2013-01-31 19:57:51   
106         Kierankane1  535909587 2013-01-31 20:08:03   
118        TheWikiPikey  535906243 2013-01-31 19:45:29   
119        TheWikiPikey  535906243 2013-01-31 19:45:29   
152             SPSpyro  535892970 2013-01-31 18:09:30   
158            Zjslayer  535785124 2013-01-31 01:03:06   
159            Zjslayer  535785272 2013-01-31 01:04:10   
160            Zjslayer  535785585 2013-01-31 01:06:08   
161            Zjslayer  535863724 2013-01-31 14:30:21   
176               Slesv  535870716 2013-01-31 15:28:57   
186           Ashman741  535857592 2013-01-31 13:32:05   
268        Drloosen2011  535781659 2013-01-31 00:35:00   
271        Drloosen2011  535782080 2013-01-31 00:38:33   
322            Jezzy68r  535739685 2013-01-30 19:51:27   
329            Wasahobo  534550640 2013-01-23 21:16:55   
330            Wasahobo  534551424 2013-01-23 21:21:36   
368             Ggg2013  535697462 2013-01-30 15:04:23   
418       Jaimenandnick  535591554 2013-01-29 22:55:45   
472         Skateman988  535553377 2013-01-29 18:35:27   
477         Skateman988  535554189 2013-01-29 18:40:47   
584        Delaware2k13  535444568 2013-01-29 02:03:47   
661             Katetha  534349763 2013-01-22 16:45:20   
662             Katetha  534349810 2013-01-22 16:45:45   
663             Katetha  534349810 2013-01-22 16:45:45   
670             Katetha  534352812 2013-01-22 17:05:54   
...                 ...        ...                 ...   
188324   Bizzenboom1234  521288062 2012-11-04 01:15:37   
188325   Bizzenboom1234  521288439 2012-11-04 01:19:00   
188326   Bizzenboom1234  521288565 2012-11-04 01:20:10   
188359   Bizzenboom1234  523397134 2012-11-16 22:16:34   
188360   Bizzenboom1234  523397459 2012-11-16 22:18:28   
188361   Bizzenboom1234  523397459 2012-11-16 22:18:28   
188362   Bizzenboom1234  524325976 2012-11-22 09:16:09   
188363   Bizzenboom1234  524325976 2012-11-22 09:16:09   
188364   Bizzenboom1234  524326242 2012-11-22 09:19:21   
188365   Bizzenboom1234  524326242 2012-11-22 09:19:21   
188366   Bizzenboom1234  524326643 2012-11-22 09:23:45   
188367   Bizzenboom1234  524326643 2012-11-22 09:23:45   
188368   Bizzenboom1234  524326778 2012-11-22 09:25:14   
188369   Bizzenboom1234  524326778 2012-11-22 09:25:14   
188370   Bizzenboom1234  524457917 2012-11-23 05:23:12   
188371   Bizzenboom1234  524457917 2012-11-23 05:23:12   
188372   Bizzenboom1234  524458467 2012-11-23 05:29:04   
188373   Bizzenboom1234  524458467 2012-11-23 05:29:04   
188374   Bizzenboom1234  527449413 2012-12-11 01:21:31   
188375   Bizzenboom1234  527449413 2012-12-11 01:21:31   
188377   Bizzenboom1234  529336756 2012-12-22 21:01:15   
188403   Bizzenboom1234  547905388 2013-03-31 00:36:14   
188404   Bizzenboom1234  564685687 2013-07-17 18:32:05   
188405   Bizzenboom1234  564685687 2013-07-17 18:32:05   
188406   Bizzenboom1234  564685977 2013-07-17 18:34:32   
188407   Bizzenboom1234  564685977 2013-07-17 18:34:32   
188448   Bizzenboom1234  591321595 2014-01-18 21:02:33   
188488         Klintrin  617665644 2014-07-20 04:37:28   
188489         Klintrin  617665831 2014-07-20 04:40:41   
188490         Klintrin  617665839 2014-07-20 04:40:48   

                                           pagetitle  isReverted  \
3                                      Pueblo Revolt        True   
16                             Ohio State University        True   
18                             Ohio State University        True   
26                                            Engine        True   
97                 Template:Green Bay Packers roster        True   
100       

In [973]:
ben_final['FirstMeta'] = ben_final.groupby('userid')['pagetitle'].first().str.contains('/')

In [974]:
len(ben_final['userid'].unique())

16496

In [975]:
van_final['FirstMeta'] = van_final.groupby('userid')['pagetitle'].first().str.contains('/')

In [976]:
np.sum(van_true)

350

In [977]:
len(van_final['userid'].unique())

17015

## from above analysis, 1533 unique benign users first edit was a metapage, and 350 vandal users first edit was a meta-page. The weighted average ration being 81:19 (benign:vandal). It implies, This implements the feature "fm" in the paper, Which states, if the first edit of the user is a meta-page, then he is more likely to be a benign user

In [978]:
#ben_final = ben_final[0:100000]
#van_final = van_final[0:100000]


## Taking the Date time stamp into account, we are analysing time difference between the adjacent edits by grouping them according to "pagetitle" and "userid" This will be useful to study the following features:
###  --> Vandals make faster edits than benign users
### -->  Vandals spend less time editing a new page
### -->  Benign users are likely to take longer to edit a new page than a vandal

In [979]:
ben_final['diffs'] = ben_final.groupby(['userid'])['revtime'].transform(lambda x: x.diff()) / np.timedelta64(1, 'm')

In [980]:
ben_final['diffs'] = ben_final['diffs'].fillna(0)

In [981]:
ben_dummy = ben_final.loc[:,['diffs','userid','pagetitle']]

In [982]:
ben_dummy.head(20)

diffs    userid                                pagetitle
0        0.000000  18342896                            User:Swiftscw
1        1.200000  18342896                            User:Swiftscw
2        5.150000  18342896                            User:Swiftscw
3       23.383333  18342896             User:Swiftscw/AMAPOLA CABASE
4        7.833333  18342896        Wikipedia:Requests for undeletion
5        0.000000  18342896        Wikipedia:Requests for undeletion
6       45.766667  18342896                             Maria Cabase
7       19.933333  18342896                             Maria Cabase
8        7.650000  18342896                             Maria Cabase
9       13.433333  18342896                             Maria Cabase
10      52.566667  18342896                             Maria Cabase
11       5.366667  18342896                             Maria Cabase
12       7.900000  18342896                        Talk:Maria Cabase
13       2.433333  18342896                            User:Swiftscw
14      12.383333  18342896                    Manuel Kabajar Cabase
15       0.666667  18342896                    Manuel Kabajar Cabase
16      50.833333  18342896              User talk:Postdlf/Archive23
17    1149.083333  18342896                        Talk:Maria Cabase
18  246173.133333  18342896  Wikipedia:WikiProject Biography/Members
19      65.166667  18342896                       User talk:Swiftscw

In [983]:
ben_dummy.groupby(['userid'])['diffs'].mean()

userid
18177826       778.140698
18178056     65067.585185
18178102     13150.735606
18178134     44033.255556
18178244         1.161905
18178257         2.733333
18178679      7556.299206
18178872    196143.520833
18178885      1168.585616
18178917      1035.111905
18178918         1.354167
18179104        83.692424
18179286         3.600000
18179499     52468.865278
18179682       656.505556
18179852         2.466667
18180185    102290.213889
18180236         7.001899
18180345      7517.307738
18180383       355.311667
18180784       236.631013
18180919       240.799621
18181044         0.000000
18181152         0.000000
18181237         2.497727
18181335      4241.415988
18181530     83914.450000
18181773         0.341667
18181786         1.002941
18181935      1599.383333
                ...      
21965844       461.264103
21966099      2772.362179
21966197      3886.234444
21966297       427.898725
21966358     43413.051667
21967185       478.888889
21967464         2.135897
21968

In [984]:
ben_dummy

diffs    userid  \
0            0.000000  18342896   
1            1.200000  18342896   
2            5.150000  18342896   
3           23.383333  18342896   
4            7.833333  18342896   
5            0.000000  18342896   
6           45.766667  18342896   
7           19.933333  18342896   
8            7.650000  18342896   
9           13.433333  18342896   
10          52.566667  18342896   
11           5.366667  18342896   
12           7.900000  18342896   
13           2.433333  18342896   
14          12.383333  18342896   
15           0.666667  18342896   
16          50.833333  18342896   
17        1149.083333  18342896   
18      246173.133333  18342896   
19          65.166667  18342896   
20       11600.750000  18342896   
21          26.150000  18342896   
22       32616.250000  18342896   
23           2.066667  18342896   
24           2.216667  18342896   
25           0.000000  18359941   
26         982.066667  18359941   
27        5986.750000  18359941   
28           0.000000  18359941   
29          65.200000  18359941   
...               ...       ...   
657833       3.266667  21807483   
657834       0.550000  21807483   
657835   49006.500000  21807483   
657836       1.616667  21807483   
657837  328630.416667  21807483   
657838     699.716667  21807483   
657839    8545.000000  21807483   
657840   26312.816667  21807483   
657841       3.300000  21807483   
657842       0.000000  21940037   
657843    1100.683333  21940037   
657844       0.000000  21932219   
657845       0.916667  21932219   
657846       1.000000  21932219   
657847       0.450000  21932219   
657848       1.900000  21932219   
657849       0.733333  21932219   
657850   10822.483333  21932219   
657851       6.716667  21932219   
657852    2544.350000  21932219   
657853       2.983333  21932219   
657854       9.066667  21932219   
657855    7052.933333  21932219   
657856    4717.950000  21932219   
657857       0.016667  21932219   
657858       5.283333  21932219   
657859       0.000000  21952946   
657860       5.383333  21952946   
657861       0.800000  21952946   
657862       1.000000  21952946   

                                                pagetitle  
0                                           User:Swiftscw  
1                                           User:Swiftscw  
2                                           User:Swiftscw  
3                            User:Swiftscw/AMAPOLA CABASE  
4                       Wikipedia:Requests for undeletion  
5                       Wikipedia:Requests for undeletion  
6                                            Maria Cabase  
7                                            Maria Cabase  
8                                            Maria Cabase  
9                                            Maria Cabase  
10                                           Maria Cabase  
11                                           Maria Cabase  
12                                      Talk:Maria Cabase  
13                                          User:Swiftscw  
14                                  Manuel Kabajar Cabase  
15                                  Manuel Kabajar Cabase  
16                            User talk:Postdlf/Archive23  
17                                      Talk:Maria Cabase  
18                Wikipedia:WikiProject Biography/Members  
19                                     User talk:Swiftscw  
20                                           Maria Cabase  
21                                           Maria Cabase  
22                                         Bobby Enriquez  
23                                         Bobby Enriquez  
24                                         Bobby Enriquez  
25                                             User:Azkle  
26      Wikipedia:Training/For students/Training feedback  
27                                                Netflix  
28                                                Netflix  
29                                           

###### checking for beningn users

In [985]:
ben_fin = ben_dummy.groupby('userid').nth(1)

In [986]:
ben_fin.shape

(16085, 2)

In [987]:
ben_fin['EditTime>15'] = ben_fin['diffs']>15

### There are 5639 benign users who take more than 15 min to edit a new page

###### checking for vandal users

In [988]:
van_final['diffs'] = van_final.groupby(['userid'])['revtime'].transform(lambda x: x.diff()) / np.timedelta64(1, 'm')

In [989]:
van_final['diffs'] = van_final['diffs'].fillna(0)

In [990]:
van_dummy = van_final.loc[:,['diffs','userid','pagetitle']]

In [991]:
van15_fin = van_dummy.groupby('userid').nth(1)

In [992]:
van15_fin['EditTime>15'] = van15_fin['diffs']>15

### There are 1951 vandal users who take more than 15min to edit a new page. This is an indication of the behavior for a feature that, beningn users take more time to edit a new page than a vandal user

In [999]:
ben_fin["stiki_mean"] = ben_scores.groupby('userid')['stiki_score'].mean()

In [1000]:
ben_fin.describe()

diffs    stiki_mean
count  1.608500e+04  16085.000000
mean   9.462651e+03      0.053216
std    5.189957e+04      0.067131
min    0.000000e+00      0.000000
25%    1.616667e+00      0.000000
50%    6.000000e+00      0.032994
75%    4.176667e+01      0.081396
max    1.328026e+06      0.632488

### checking for the behavior where, the average of the stikiscore given to user for the edits that has been made, crosses a threshold value

In [1003]:
ben_fin['stiki_percent']=ben_fin['stiki_mean']<0.1

In [1004]:
ben_fin['stiki_percent'].value_counts()

True     13019
False     3066
Name: stiki_percent, dtype: int64

In [1001]:
van15_fin["stiki_mean"] = van_scores.groupby('userid')['stiki_score'].mean()

In [1002]:
van15_fin.describe()

diffs    stiki_mean
count  1.619200e+04  16192.000000
mean   5.641306e+03      0.267262
std    7.712583e+04      0.176928
min    0.000000e+00      0.000000
25%    0.000000e+00      0.139955
50%    1.600000e+00      0.260055
75%    4.900000e+00      0.380814
max    3.450404e+06      0.929759

In [1005]:
van15_fin['stiki_percent']=van15_fin['stiki_mean']<0.1

In [1006]:
van15_fin['stiki_percent'].value_counts()

False    13063
True      3129
Name: stiki_percent, dtype: int64

In [1007]:
van15_fin.head()

diffs                                pagetitle  EditTime>15  \
userid                                                                   
107645  3.300000               Wikipedia:Community portal        False   
284493  1.133333  Wikipedia:Articles for deletion/Urban75        False   
301475  7.166667                               Dick Allen        False   
340857  0.016667                                  Rolling        False   
345175  3.300000                   Surrey Central Station        False   

        stiki_mean  stiki_percent  
userid                             
107645    0.000000           True  
284493    0.001028           True  
301475    0.000000           True  
340857    0.000000           True  
345175    0.000000           True

In [1028]:
ben_final['isReverted'].value_counts()

False    612657
True      45206
Name: isReverted, dtype: int64

In [1029]:
van_final['isReverted'].value_counts()

True     136499
False     51992
Name: isReverted, dtype: int64

In [1010]:
ben_fin['FirstMeta'] = ben_final.groupby('userid')['pagetitle'].first().str.contains('/')

In [1011]:
van15_fin['FirstMeta'] = van_final.groupby('userid')['pagetitle'].first().str.contains('/')

In [1012]:
ben_fin.head()

diffs                   pagetitle  EditTime>15  stiki_mean  \
userid                                                                      
18177826  502.500000  User:Shamprabhakar/sandbox         True    0.000000   
18178056   78.850000           User:Jack Odanaka         True    0.034222   
18178102    9.800000          User:DanielQueiros        False    0.036366   
18178134    2.533333            User talk:Arre 9        False    0.000000   
18178244    1.683333               Lee Chong Wei        False    0.000000   

          stiki_percent  FirstMeta  
userid                              
18177826           True      False  
18178056           True      False  
18178102           True      False  
18178134           True      False  
18178244           True      False

### checking for a behaviour where the number of times the edit is reverted by cluebot is more frequent. 

In [1030]:
from scipy import stats
ben_fin['reverted_mode'] = ben_final.groupby(['userid']).agg({'isReverted': lambda x:stats.mode(x)[0]})

In [1031]:
van15_fin['reverted_mode'] = van_final.groupby(['userid']).agg({'isReverted': lambda x:stats.mode(x)[0]})

In [1032]:
ben_fin['reverted_mode'].value_counts()

False    15130
True       955
Name: reverted_mode, dtype: int64

In [1033]:
van15_fin['reverted_mode'].value_counts()

True     14918
False     1274
Name: reverted_mode, dtype: int64

In [1034]:
ben_fin.head()

diffs                   pagetitle  EditTime>15  stiki_mean  \
userid                                                                      
18177826  502.500000  User:Shamprabhakar/sandbox         True    0.000000   
18178056   78.850000           User:Jack Odanaka         True    0.034222   
18178102    9.800000          User:DanielQueiros        False    0.036366   
18178134    2.533333            User talk:Arre 9        False    0.000000   
18178244    1.683333               Lee Chong Wei        False    0.000000   

          stiki_percent  FirstMeta  cluebot_mode  reverted_mode  
userid                                                           
18177826           True      False             0           True  
18178056           True      False             0          False  
18178102           True      False             0          False  
18178134           True      False             0          False  
18178244           True      False             0           True

In [1035]:
ben_out = ben_fin.drop(['diffs','pagetitle','stiki_mean','cluebot_mode'],axis=1)

In [1036]:
van_out = van15_fin.drop(['diffs','pagetitle','stiki_mean','cluebot_mode'],axis=1)

In [1037]:
ben_out.to_csv('benign_out.csv')
van_out.to_csv('vandal_out.csv')

In [1038]:
ben_out['type']='Benign'
van_out['type']='Vandal'

In [1039]:
data_final = ben_out.append(van_out)

## This is the final data set that we have build based on the analysis of behaviour depicted by the user while edit a page. This will be integrated into the model to make the vandal prediction for test data.

In [1040]:
data_final.head()

EditTime>15  stiki_percent  FirstMeta  reverted_mode    type
userid                                                                
18177826         True           True      False           True  Benign
18178056         True           True      False          False  Benign
18178102        False           True      False          False  Benign
18178134        False           True      False          False  Benign
18178244        False           True      False           True  Benign

In [1041]:
data_final.to_csv('Data_final.csv')

In [1043]:
data_final.columns

Index(['EditTime>15', 'stiki_percent', 'FirstMeta', 'reverted_mode', 'type'], dtype='object')

In [1045]:
X = data_final[['EditTime>15', 'stiki_percent', 'FirstMeta', 'reverted_mode']]
y = data_final['type']

### From Scikit-Learn Importing train_test_split function
### Splitting input and target data into train and test data

In [1046]:
from sklearn.model_selection import train_test_split

In [1047]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

### From Scikit-Learn importing Logistic Regression Function

In [1048]:
from sklearn.linear_model import LogisticRegression

### Building Logistic Regression Model

In [1049]:
lr = LogisticRegression()
lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

### Predicting the Target Variable using the model build above

In [1050]:
ypred = lr.predict(X_test)

### From Scikit-Learn Importing Confusion_matrix and Accuracy

In [1051]:
from sklearn.metrics import confusion_matrix,accuracy_score

### Comparing Actual and Predicted Target values using Confusion_Matrix

In [1052]:
confusion_matrix(y_test,ypred)

array([[5013,  337],
       [ 415, 4887]])

### Calculating Accuracy

In [1053]:
accuracy_score(y_test,ypred)

0.92940292902741273